File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/VoenmehScheduleClient.kt
package ru.bgtu_voenmeh.zapara.data

import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit

class VoenmehScheduleClient(
    private val get: (String) -> String = Companion::httpGet
) {
    suspend fun fetchSchedule(groupNames: Collection<String> = emptyList()): ParsedSchedule = coroutineScope {
        val meta = VoenmehScheduleParser.parseMeta(get(META_URL))
        val fetchList = if (groupNames.isEmpty()) {
            meta.groups
        } else {
            groupNames.map { it.trim() }.filter { it.isNotEmpty() }.distinct()
        }
        val slots = Semaphore(6)
        val loaded = fetchList.map { name ->
            async {
                slots.withPermit {
                    val url = "$LESSONS_URL?name=${encode(name)}&type=group"
                    name to try {
                        VoenmehScheduleParser.parseLessons(get(url), name)
                    } catch (_: Exception) {
                        emptyList()
                    }
                }
            }
        }.awaitAll()
        VoenmehScheduleParser.assemble(meta, loaded)
    }

    companion object {
        const val ORIGIN = "https://voenmeh.ru/schedule"
        const val META_URL = "https://voenmeh.ru/api/schedule/meta"
        const val LESSONS_URL = "https://voenmeh.ru/api/schedule/lessons"

        fun encode(value: String): String =
            URLEncoder.encode(value, StandardCharsets.UTF_8.name()).replace("+", "%20")

        fun httpGet(url: String): String {
            val conn = URL(url).openConnection() as HttpURLConnection
            try {
                conn.setRequestProperty("User-Agent", "Mozilla/5.0 (Linux; Android) Zapara/1.0")
                conn.setRequestProperty("Accept", "application/json")
                conn.connectTimeout = 20_000
                conn.readTimeout = 30_000
                conn.connect()
                val stream = if (conn.responseCode in 200..299) conn.inputStream else conn.errorStream
                val body = stream?.bufferedReader(Charsets.UTF_8)?.readText().orEmpty()
                if (conn.responseCode == HttpURLConnection.HTTP_NOT_FOUND) return """{"lessons":[]}"""
                if (conn.responseCode != HttpURLConnection.HTTP_OK) {
                    throw IllegalStateException("HTTP ${conn.responseCode}")
                }
                return body
            } finally {
                conn.disconnect()
            }
        }
    }
}

File: src/Vograph.Timetable/VoenmehScheduleClient.cs
using Vograph.Core.Models;

namespace Vograph.Timetable;

public sealed class VoenmehScheduleClient
{
    public const string Origin = "https://voenmeh.ru/schedule";
    public const string MetaUrl = "https://voenmeh.ru/api/schedule/meta";
    public const string LessonsUrl = "https://voenmeh.ru/api/schedule/lessons";

    private readonly Func<string, CancellationToken, Task<string>> _get;

    public VoenmehScheduleClient(Func<string, CancellationToken, Task<string>> get) => _get = get;

    public Task<ParsedSchedule> FetchAsync(IReadOnlyList<string> groupNames, CancellationToken ct = default)
        => FetchAsync(groupNames, meta: null, ct);

    public async Task<ParsedSchedule> FetchAsync(IReadOnlyList<string> groupNames, VoenmehScheduleMeta? meta, CancellationToken ct)
    {
        meta ??= VoenmehScheduleParser.ParseMeta(await _get(MetaUrl, ct).ConfigureAwait(false));
        var fetchList = groupNames.Select(n => n.Trim()).Where(n => n.Length > 0).Distinct(StringComparer.Ordinal).ToList();
        var loaded = new List<(string Name, List<Lesson> Lessons)>(fetchList.Count);
        using var gate = new SemaphoreSlim(6);
        var tasks = fetchList.Select(async name =>
        {
            await gate.WaitAsync(ct).ConfigureAwait(false);
            try
            {
                var url = $"{LessonsUrl}?name={Uri.EscapeDataString(name)}&type=group";
                var lessons = VoenmehScheduleParser.ParseLessons(await _get(url, ct).ConfigureAwait(false), name);
                return (name, lessons);
            }
            finally { gate.Release(); }
        });
        loaded.AddRange(await Task.WhenAll(tasks).ConfigureAwait(false));
        return VoenmehScheduleParser.Assemble(meta, loaded);
    }
}
